### PR TITLE
Switch from cl100k_base to gpt2 for Improved Subword Tokenization

### PR DESCRIPTION
**Pull request by klei22**
# Switch from cl100k_base to gpt2 for Improved Subword Tokenization

## Changes Made
- Changed tiktoken encoding from `cl100k_base` to `gpt2` in tests and prepare.py
- Updated tokenization logic in tests to handle subword tokens properly
- Improved token boundary visualization in test output
- Maintained proper handling of SNAC special tokens

## Benefits
1. **Better Subword Tokenization**: The `gpt2` encoding provides more natural subword tokenization compared to `cl100k_base`. For example:
   - "Here" → single token (4342)
   - "is" → single token with space (318)
   - "and" → single token with space (290)
   - "token" → single token with space (11241)
   - "test" → single token with space (1332)

2. **Special Token Support**: Maintains proper handling of SNAC tokens:
   - `<snac>4097</snac>` → single token (100000)
   - `<snac>100</snac>` → single token (100001)

3. **Improved Visualization**: Enhanced token boundary display in tests for better debugging and understanding:
   ```
   |Here| is| a| |<snac>4097</snac>| and| a| |<snac>100</snac>| token| test|.|
   ```

## Testing
- All tests pass with the new encoding
- Token boundaries are properly preserved
- Special tokens are correctly handled
- Metadata is properly saved and loaded
- Token counting functionality works as expected

## Next Steps
- [ ] Consider implementing a custom tokenizer solution for even more control over tokenization
- [ ] Add more comprehensive tests for edge cases
- [ ] Document the tokenization behavior differences between encodings

## Testing Instructions
To verify the changes:
1. Run the tests: `python tests.py`
2. Check the token boundary visualization in the test output
3. Verify that both regular text and SNAC tokens are properly handled

### CODE REVIEW

File: data/template/prepare.py [hunk 1]
```diff
@@ -114,15 +86,28 @@ def save_tokens(ids, output_file, dtype):
...
+        with open("meta.pkl", "rb") as f:
+            meta = pickle.load(f)
+        meta.update({
+            "has_additional_tokens": True,
+            "special_tokens": additional_tokens,
+            "tokenizer": "tiktoken",
+            "tiktoken_encoding": args.tiktoken_encoding
+        })
+        with open("meta.pkl", "wb") as f:
+            pickle.dump(meta, f)
```
Comment (gkielian) [inline data/template/prepare.py:110]: Hmm, I think we should move this to the tokenizers.py file within the Tiktoken tokenizer class

File: data/template/prepare.py [hunk 2]
```diff
@@ -13,74 +13,46 @@
...
     # Char only arguments
     parser.add_argument("--reuse_chars", action="store_true", help="Reuse character list")
     # Add argument for custom characters file
     parser.add_argument("--custom_chars_file", type=str, default=None, help="Path to the file containing custom characters for the tokenizer")
-    # Customize output names for bins
-    parser.add_argument("--train_output", type=str, default="train.bin", help="Output file for tokenized training data")
-    parser.add_argument("--val_output", type=str, default="val.bin", help="Output file for tokenized validation data")
-    # Options for using separate training and validation input files
-    parser.add_argument("-s", "--use_separate_files", action="store_true", help="Use separate files for training and validation input")
-    parser.add_argument("-t", "--train_input", type=str, help="Path to the training input text file")
```
Comment (gkielian) [inline data/template/prepare.py:37]: I think we'll have to retain the --train_input, --percentage_train, --json_tokens_file.

I think we can remove the numeric range tokenization, as we can create a separate script to generate a json_tokens file for this.

I don't think use separate files is used, but we should check the data folder first. (and the demos folder too) before merging.

File: data/template/prepare.py [hunk 3]
```diff
@@ -114,15 +86,28 @@ def save_tokens(ids, output_file, dtype):
                 batch = ids[i:i+batch_size]
                 np.array(batch, dtype=dtype).tofile(f_out)
 
-    if (args.method == "tiktoken" and args.tiktoken_encoding == "cl100k_base") or (args.method == "numeric_range" and args.max_token > 65535):
-        dtype = np.uint32
-    else:
-        dtype = np.uint16
+    # Determine dtype based on token IDs
+    max_token_id = max(max(train_ids), max(val_ids) if val_ids else 0)
+    dtype = np.uint32 if max_token_id > 65535 else np.uint16
```
Comment (gkielian) [inline data/template/prepare.py:91]: This is great!

**Review by gkielian — CHANGES_REQUESTED** [review]

**Comment by gkielian** [comment]
Btw, edited the title and changelist, to trim to the major changes in the commits.
